scIB/metrics/kbet.py

```diff
@@ -0,0 +1,167 @@
+import numpy as np
+import scipy.sparse
+import pandas as pd
+import logging
+import rpy2.robjects as ro
+import rpy2.rinterface_lib.callbacks
+import anndata2ri
+import scanpy as sc
+
+from scIB.utils import checkAdata, checkBatch
+from .utils import diffusion_conn, diffusion_nn
+
+rpy2.rinterface_lib.callbacks.logger.setLevel(logging.ERROR)  # Ignore R warning messages
+
+
+def kBET_single(
+        matrix,
+        batch,
+        type_=None,
+        k0=10,
+        knn=None,
+        subsample=0.5,
+        heuristic=True,
+        verbose=False
+):
+    """
+    params:
+        matrix: expression matrix (at the moment: a PCA matrix, so do.pca is set to FALSE
+        batch: series or list of batch assignemnts
+        subsample: fraction to be subsampled. No subsampling if `subsample=None`
+    returns:
+        kBET p-value
+    """
+
+    anndata2ri.activate()
+    ro.r("library(kBET)")
+
+    if verbose:
+        print("importing expression matrix")
+    ro.globalenv['data_mtrx'] = matrix
+    ro.globalenv['batch'] = batch
+    # print(matrix.shape)
+    # print(len(batch))
+
+    if verbose:
+        print("kBET estimation")
+    # k0 = len(batch) if len(batch) < 50 else 'NULL'
+
+    ro.globalenv['knn_graph'] = knn
+    ro.globalenv['k0'] = k0
+    batch_estimate = ro.r(
+        f"batch.estimate <- kBET(data_mtrx, batch, knn=knn_graph, k0=k0, plot=FALSE, do.pca=FALSE, heuristic=FALSE, adapt=FALSE, verbose={str(verbose).upper()})")
+
+    anndata2ri.deactivate()
+    try:
+        ro.r("batch.estimate$average.pval")[0]
```

Comment:
Why do you use the average p-value, not the observed rejection rate?

Reply:
I don't quite remember what we agreed upon, this is just the code we'd been using so far.

Reply:
Well, we *should* be using the observed rejection rate, because this is what has been evaluated in the kBET paper, not the p-values. The observed rejection rate is between 0 and 1, with 0 being best and 1 poorest possible. Low observed rejection rates are correlated with high p-values, so the metric might be mis-scaled.

Reply:
If I'm not mistaken, this is your code from 2 years ago @mbuttner. How would you change it? If this is incorrect, we should remove kBET from the evaluation very soon.

Reply:
Well, Michaela wrote the first skeleton of the kBET metric for scIB and I did not realise that the p-values were used instead of the observed rejection rate. I should have found this earlier, I know.

Reply:
#257